### PR TITLE
Call hierarchy support

### PR DESCRIPTION
**Pull request by July541**
Hello all,

This pr is about my GSoC 2021 project, call hierarchy support. I create a draft to show my progress to those who may be interested.

**Progress**

- [x] Prepare call hierarchy (working now)
- [x] Incoming calls
- [x] Outgoing calls

**Known issues**

- [x] Data family range error (see commented test)
- [x] Incoming/Outgoing test missing
- [x] Name display not optimized for `DuplicateRecordFields`

### CODE REVIEW

**Comment by wz1000** [comment]
If I understand correctly, call hierarchy is meant to be global and project wide. However, you seem to be generating the tree using a single `HieAst`.

I think it would be a good idea to make use of `hiedb` for this rather than directly using `.hie` files/`HieAst`. Take a look at the `getGraph`, `getReachable` and `getVertices` functions in hiedb: https://github.com/wz1000/HieDb/blob/master/src/HieDb/Query.hs#L175

The code you have for traversing the `HieAst` could still be very useful - we only keep up to date version of files the user isn't currently editing in the `hiedb`. For the files the user has open in their editor (`IsFileOfInterest => True`), to get an accurate callgraph the information from the `HieAst` traversal would need to be combined with the information from `hiedb`.

**Comment by July541** [comment]
> If I understand correctly, call hierarchy is meant to be global and project wide. However, you seem to be generating the tree using a single `HieAst`.

You're right, currently, I just finished the part of `prepare call hierarchy`. This part only constructs the `CallHierarchyItem` without any relation to other statements. I think `HieAst` is enough now.

The `hierarchy` is queried at the `incoming/outgoing` steps, which I have not started, your suggestion is very feasible, I'll follow it in the next part.

**Comment by pepeiborra** [comment]
@July541 I can see that you uploaded some commits yesterday. Care to share an update on what the recent changes are?

**Comment by July541** [comment]
@pepeiborra It's about a prototype of the implementation of outgoing calls which has functionality but a bad code style. I'll show a complete call hierarchy after incoming calls are finished, it'll come soon.

**Comment by July541** [comment]
https://user-images.githubusercontent.com/16057632/125197743-7e2dd400-e291-11eb-8a48-872dda05e900.mp4

Recent progress, incoming calls and outgoing calls have basic functions, see demo above for details.

**Current problems**:
- [ ] Duplication items(multi range merge)
- [ ] Range of incoming/outgoing calls still exist problems
- [ ] Test missing
- [ ] Query will be broken if the target is a parameter

**Comment by pepeiborra** [comment]
This looks quite reasonable to me, nice job @July541 ! 
I would argue for landing it and then opening new tickets to fix the outstanding problems. 
What do you think ?

File: plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import           GHC
+import           HieDb                          (HieDb (getConn), Symbol (..),
+                                                 toNsChar)
+import qualified HieDb
+import           Ide.Plugin.CallHierarchy.Types
+import           Module
+import           Name
+
+incomingCalls :: HieDb -> Symbol -> IO [Vertex]
+incomingCalls (getConn -> conn) symbol = do
```
Comment (wz1000) [inline plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs:17]: It would be best to move these functions into `hiedb`.

**Comment by July541** [comment]
@pepeiborra I agree with you. I think it can be reviewed after the break issue is fixed, not take long.

@wz1000 It turns out I wrote it in `HieDb` from the beginning. But the following questions bothered me and made me turn to the internal implementation:

1. The definition of `Vertex` is vague, we may need to rewrite the relevant parts, in which the compatibility comes at a high price to keep code elegant(for `HieDb` developer, not user).

2. The two calls(incoming/outgoing) seem to be domain-specific, and I'm not sure they will have other usages, so I wrote them inside the call hierarchy.

Honestly, I'm not sure it's a reasonable solution. if you think that move to `HieDb` is better, I'll transfer it.

**Comment by pepeiborra** [comment]
> @pepeiborra I agree with you. I think it can be reviewed after the break issue is fixed, not take long.
> 

Awesome, please shout if you need any help

**Comment by isovector** [comment]
Your CI appears to be blocked by https://github.com/haskell/haskell-language-server/issues/2016

File: stack-8.10.4.yaml [hunk 2]
```diff
@@ -53,16 +54,21 @@ extra-deps:
...
-  - lsp-test-0.14.0.0
   - dependent-map-0.4.0.0@sha256:ca2b131046f4340a1c35d138c5a003fe4a5be96b14efc26291ed35fd08c62221,1657
   - dependent-sum-0.7.1.0@sha256:5599aa89637db434431b1dd3fa7c34bc3d565ee44f0519bfbc877be1927c2531,2068
   - dependent-sum-template-0.1.0.3@sha256:0bbbacdfbd3abf2a15aaf0cf2c27e5bdd159b519441fec39e1e6f2f54424adde,1682
   - constraints-extras-0.3.0.2@sha256:013b8d0392582c6ca068e226718a4fe8be8e22321cc0634f6115505bf377ad26,1853
   - some-1.0.1@sha256:26e5bab7276f48b25ea8660d3fd1166c0f20fd497dac879a40f408e23211f93e,2055
   - unliftio-core-0.2.0.1@sha256:9b3e44ea9aacacbfc35b3b54015af450091916ac3618a41868ebf6546977659a,1082
 
+  - github: July541/lsp
+    commit: ef59c28b41ed4c5775f0ab0c1e985839359cec96
```
Comment (pepeiborra) [inline stack-8.10.4.yaml:65]: Can we avoid pointing to a commit in your fork, and instead point to a commit in `haskell/lsp`?

**Review by pepeiborra — CHANGES_REQUESTED** [review]

File: stack-8.10.4.yaml (hunk 2 above)
Comment (July541) [reply stack-8.10.4.yaml]: Sure

**Comment by pepeiborra** [comment]
@July541 You need to rebase this on master and fix the conflicts and/or test failures. Is there anything blocking you?

**Comment by July541** [comment]
@pepeiborra Yes, I am working on FOI queries, it now works as expected. But I don't know why the test hanging so long.

**Comment by pepeiborra** [comment]
> @pepeiborra Yes, I am working on FOI queries, it doesn't work as expected. But I don't know why the test hanging so long.

Our tests are written as parsers for LSP sessions. Most of the failure modes are timeouts while waiting for an expected LSP message. Is that what you mean?

**Comment by July541** [comment]
> > @pepeiborra Yes, I am working on FOI queries, it doesn't work as expected. But I don't know why the test hanging so long.
> 
> Our tests are written as parsers for LSP sessions. Most of the failure modes are timeouts while waiting for an expected LSP message. Is that what you mean?

Maybe not, I mean the 28 waiting tests listed below my pr.

**Comment by pepeiborra** [comment]
> > > @pepeiborra Yes, I am working on FOI queries, it doesn't work as expected. But I don't know why the test hanging so long.
> > 
> > 
> > Our tests are written as parsers for LSP sessions. Most of the failure modes are timeouts while waiting for an expected LSP message. Is that what you mean?
> 
> Maybe not, I mean the 28 waiting tests listed below my pr.

I believe that's because you need to resolve the conflicts, otherwise the Github Actions don't run

**Review by pepeiborra — APPROVED** [review]

**Comment by July541** [comment]
> I believe that's because you need to resolve the conflicts, otherwise the Github Actions don't run

Got it!

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -201,6 +201,6 @@ jobs:
         name: Test hls-refine-imports-plugin test suite
         run: cabal test hls-refine-imports-plugin --test-options="-j1 --rerun-update" || cabal test hls-refine-imports-plugin --test-options="-j1 --rerun" || LSP_TEST_LOG_COLOR=0 LSP_TEST_LOG_MESSAGES=true LSP_TEST_LOG_STDERR=true cabal test hls-refine-imports-plugin --test-options="-j1 --rerun"
 
-      - if: ${{ needs.pre_job.outputs.should_skip != 'true' && matrix.test && matrix.ghc != '9.0.1' }}
+      - if: ${{ needs.pre_job.outputs.should_skip != 'true' && matrix.test && matrix.ghc }}
```
Comment (jneira) [inline .github/workflows/test.yml:204]:
```suggestion
      - if: ${{ needs.pre_job.outputs.should_skip != 'true' && matrix.test}}
```

File: plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs [hunk 4]
```diff
@@ -19,55 +20,75 @@ import           Name
...
+        u = unitString $ moduleUnit symModule
+    in  (o, m, u)
+#else
+parseSymbol :: Symbol -> (String, String, String)
+parseSymbol Symbol{..} =
+    let o = toNsChar (occNameSpace symName) : occNameString symName
+        m = moduleNameString $ moduleName symModule
+        u = unitIdString $ moduleUnitId symModule
+    in  (o, m, u)
+#endif
```
Comment (pepeiborra) [inline plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs]: Avoid using CPP here, instead look to reuse one of the helpers defined in the module `Development.IDE.GHC.Compat` specifically for this purpose.

File: plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 
 module Ide.Plugin.CallHierarchy.Query (
   incomingCalls
 , outgoingCalls
 , getSymbolPosition
 ) where
 
+import qualified Data.Text                      as T
 import           Database.SQLite.Simple
 import           GHC
```
Comment (pepeiborra) [inline plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs:14]: import `Development.IDE.GHC.Compat` instead

**Review by pepeiborra — CHANGES_REQUESTED** [review]

File: plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs (hunk 4 above)
Comment (pepeiborra) [reply plugins/hls-call-hierarchy-plugin/src/Ide/Plugin/CallHierarchy/Query.hs]: The helper in this case is called simply `moduleUnitId` and it does the right thing for all the supported GHC versions

**Review by pepeiborra — APPROVED** [review]

**Comment by July541** [comment]
I think the current fail can be fixed after the new lsp is released.

**Comment by pepeiborra** [comment]
This looks ready to merge, great job @July541 !!!
Do you want to do the honours @jneira ?

**Comment by jneira** [comment]
@July541 amazing work, hls users will be amazed, congrats
And thanks you @pepeiborra for taking care, I did not have all time I wish

**Comment by jhrcek** [comment]
I pulled master this morning and I'm also not able to do `stack ./install.hs hls-8.8.4`.
I'm getting build failures similar to what I see in CI runs:
e.g. this
<details>
  <summary>build failures in CI</summary>
  <pre>
Linking .stack-work/dist/x86_64-linux/Cabal-3.0.1.0/build/ghcide/ghcide ...
ghcide                           > /root/.stack/snapshots/x86_64-linux/26f5581bf7f9f176d3f7ba334912f1d717717a1ecee407868101798f77c5c139/8.8.3/lib/x86_64-linux-ghc-8.8.3/libHSlsp-1.2.0.0-LiIFgQ1fCTv73zSCj7uRmO-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziCallHierarchy_zdfFromJSONCallHierarchyOutgoingCallsParams_closure'
ghcide                           > /root/.stack/snapshots/x86_64-linux/26f5581bf7f9f176d3f7ba334912f1d717717a1ecee407868101798f77c5c139/8.8.3/lib/x86_64-linux-ghc-8.8.3/libHSlsp-1.2.0.0-LiIFgQ1fCTv73zSCj7uRmO-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziCallHierarchy_zdfFromJSONCallHierarchyIncomingCallsParams_closure'
ghcide                           > /root/.stack/snapshots/x86_64-linux/26f5581bf7f9f176d3f7ba334912f1d717717a1ecee407868101798f77c5c139/8.8.3/lib/x86_64-linux-ghc-8.8.3/libHSlsp-1.2.0.0-LiIFgQ1fCTv73zSCj7uRmO-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziCallHierarchy_zdfFromJSONCallHierarchyPrepareParams_closure'
ghcide                           > /root/.stack/snapshots/x86_64-linux/26f5581bf7f9f176d3f7ba334912f1d717717a1ecee407868101798f77c5c139/8.8.3/lib/x86_64-linux-ghc-8.8.3/libHSlsp-1.2.0.0-LiIFgQ1fCTv73zSCj7uRmO-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziMethod_zdfFromJSONSMethod54_closure'
ghcide                           > /root/.stack/snapshots/x86_64-linux/26f5581bf7f9f176d3f7ba334912f1d717717a1ecee407868101798f77c5c139/8.8.3/lib/x86_64-linux-ghc-8.8.3/libHSlsp-1.2.0.0-LiIFgQ1fCTv73zSCj7uRmO-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziMethod_zdfFromJSONSMethod55_closure'
ghcide                           > /root/.stack/snapshots/x86_64-linux/26f5581bf7f9f176d3f7ba334912f1d717717a1ecee407868101798f77c5c139/8.8.3/lib/x86_64-linux-ghc-8.8.3/libHSlsp-1.2.0.0-LiIFgQ1fCTv73zSCj7uRmO-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziMethod_zdWSTextDocumentPrepareCallHierarchy_closure'
ghcide                           > /root/build/ghcide/.stack-work/dist/x86_64-linux/Cabal-3.0.1.0/build/libHSghcide-1.4.0.3-GzSd9eXvctl5jyqcFJIb49-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziMethod_zdfShowSMethod114_closure'
ghcide                           > /root/build/ghcide/.stack-work/dist/x86_64-linux/Cabal-3.0.1.0/build/libHSghcide-1.4.0.3-GzSd9eXvctl5jyqcFJIb49-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziMethod_zdfShowSMethod112_closure'
ghcide                           > /root/build/ghcide/.stack-work/dist/x86_64-linux/Cabal-3.0.1.0/build/libHSghcide-1.4.0.3-GzSd9eXvctl5jyqcFJIb49-ghc8.8.3.so: error: undefined reference to 'lspzmtypeszm1zi2zi0zi0zmDHDZZnuOdhmBHZZjePkaCNbQ_LanguageziLSPziTypesziMethod_zdfShowSMethod116_closure'
ghcide                           > collect2: error: ld returned 1 exit status
ghcide                           > `gcc' failed in phase `Linker'. (Exit code: 1)
  </pre>
</details>

see [failure in CI](https://app.circleci.com/pipelines/github/haskell/haskell-language-server/3411/workflows/ddb4955c-520a-4431-9a0a-8b8beda1e1c8/jobs/29193)

**Comment by pepeiborra** [comment]
Have you tried cabal clean ?

**Comment by pepeiborra** [comment]
> @pepeiborra I noticed an unexpected `ghcide` fail on 8.6.5, not sure if it is caused by me.

Some tests are flaky in Windows, that's not caused by this PR

**Comment by July541** [comment]
@jhrcek Ideally the updated dependencies are OK for the build. https://github.com/haskell/haskell-language-server/blob/d815d0474d226e00ba2e3f93f9878b109e1ea2e0/stack-8.8.4.yaml#L83-L89

**Comment by jneira** [comment]
@jhrcek i hope #2044 (which remove .stack-work from cache, so equivalent to do a `stack clean --full`) will take ride of those circleci linker errors

**Comment by pepeiborra** [comment]
> @jhrcek i hope #2044 (which remove .stack-work from cache, so equivalent to do a `stack clean --full`) will take ride of those circleci linker errors

Users will still be affected by the stack build problem though.

**Comment by jhrcek** [comment]
I tried `stack clean --full` in haskell-language-server repo, but still got build failure during linking.
In the end I removed all build snapshots from `~/.stack/snapshots/x86_64-linux-tinfo6/` and then it built correctly.
The call hierarchy seems to be working great. Thank you! :heart:

**Pull request merged**